kernel/src/process.rs

```diff
@@ -119,127 +119,137 @@ impl fmt::Debug for ProcessLoadError {
+                // we are at the end of flash, and there are no more apps to
+                // load.
+                return Ok(());
+            }
+        };
 
-            // Pass the first eight bytes to tbfheader to parse out the length
-            // of the tbf header and app. We then use those values to see if we
-            // have enough flash remaining to parse the remainder of the header.
-            let (version, header_length, entry_length) = match tbfheader::parse_tbf_header_lengths(
-                test_header_slice
-                    .try_into()
-                    .or(Err(ProcessLoadError::InternalError))?,
-            ) {
-                Ok((v, hl, al)) => (v, hl, al),
-                Err(tbfheader::InitialTbfParseError::InvalidHeader(entry_length)) => {
-                    // If we could not parse the header, then we want to skip
-                    // over this app and look for the next one.
-                    (0, 0, entry_length)
-                }
-                Err(tbfheader::InitialTbfParseError::UnableToParse) => {
-                    // Since Tock apps use a linked list, it is very possible
-                    // the header we started to parse is intentionally invalid
-                    // to signal the end of apps. This is ok and just means we
-                    // have finished loading apps.
-                    return Ok(());
-                }
-            };
+        // Pass the first eight bytes to tbfheader to parse out the length of
+        // the tbf header and app. We then use those values to see if we have
+        // enough flash remaining to parse the remainder of the header.
+        let (version, header_length, entry_length) = match tbfheader::parse_tbf_header_lengths(
+            test_header_slice
+                .try_into()
+                .or(Err(ProcessLoadError::InternalError))?,
+        ) {
+            Ok((v, hl, el)) => (v, hl, el),
+            Err(tbfheader::InitialTbfParseError::InvalidHeader(entry_length)) => {
+                // If we could not parse the header, then we want to skip over
+                // this app and look for the next one.
+                (0, 0, entry_length)
+            }
+            Err(tbfheader::InitialTbfParseError::UnableToParse) => {
+                // Since Tock apps use a linked list, it is very possible the
+                // header we started to parse is intentionally invalid to signal
+                // the end of apps. This is ok and just means we have finished
+                // loading apps.
+                return Ok(());
+            }
+        };
+
+        // Now we can get a slice which only encompasses the length of flash
+        // described by this tbf header.  We will either parse this as an actual
+        // app, or skip over this region.
+        let entry_flash = remaining_flash
+            .get(0..entry_length as usize)
+            .ok_or(ProcessLoadError::NotEnoughFlash)?;
```

Comment:
Instead of calling a `get` twice (here, and at the end of this function to update the `remaining_flash`), we can use the [slice::split_at](https://doc.rust-lang.org/std/primitive.slice.html#method.split_at) method which is more idiomatic and does the bounds check once (I see that we already use `slice::split_at_mut` in this file).

The only caveat is that it would panic on bounds checking, so we have to do a check ourselves.

```rust
// We do a manual bounds check here to return an error rather than panicking in split_at.
if entry_length > remaining_flash.len() {
    return Err(ProcessLoadError::NotEnoughFlash);
}
let (entry_flash, rest) = remaining_flash.split_at(entry_length);
remaining_flash = rest;
```

Reply:
This only runs at boot, and I'm partial to using non panic APIs.

Reply:
I agree that we need to report an error rather than panicking (as suggested in my code snippet), but apart from that `split_at` is more idiomatic, and doing twice the bounds check in separate sections of this function looks like an anti-pattern to me. With `split_at`, there's no risk of accidentally having `entry_flash` and the next `remaining_flash` overlap each other.

Reply:
If there is a split_at function that cannot panic I would agree with that being more idiomatic. But I think we need to push back against panic-able functions being idiomatic. While the compiler can omit the panic when it can verify it cannot occur, we 1) have to rely on the compiler to do that, and hope that it always can, and 2) changes to unrelated code can cause the panic to come back.

Reply:
Can we move the following lines just after these ones, so that it's clear that the split is happening at the same offset?
```rust
        // Advance the flash slice for process discovery beyond this last entry.
        // This will be the start of where we look for a new process since Tock processes
        // are allocated back-to-back in flash.
        remaining_flash = remaining_flash
            .get(entry_flash.len()..)
            .ok_or(ProcessLoadError::NotEnoughFlash)?;
```